### ISSUE TITLE
SandboxChanged  -  Pod sandbox changed, it will be killed and re-created recurring pod won't start

### ISSUE THREAD

**Issue by JJir** [Azure/acs-engine#3596]
**Is this a request for help?**:
Yes

---

**Is this an ISSUE or FEATURE REQUEST?** (choose one):
ISSUE

---

**What version of acs-engine?**:
0.19.1

---

**Orchestrator and version (e.g. Kubernetes, DC/OS, Swarm)**
v1.11.0

**What happened**:

Series of events:
Cluster created 11 days ago.
Running single pod  with single container hosting asp.net core 2.1 web server and running additional processes that run .net framework 4.6 bits.
Some clients connected. (windows 1709 based pool and image based on microsoft/dotnet-framework:4.7.2-runtime-20180615-windowsservercore-1709)
~1 days ago symbolruntime-6bdb459f-9xxnz 14:28 18/07/30 pod evicted  due to:
Status:         Failed
Reason:         Evicted
Message:        The node was low on resource: ephemeral-storage. Container symbolruntime was using 26784188Ki, which exceeds its request of 0.

The pod recreated, clients reconnected. Everything still OK at this point.

(This might be unrelated but:)
At around 0700 18/7/31 UTC a DNS Name was added to the Public IP Address
maclustermls2-a13fdb5768bef11e8b947000d3af946c  
Done using azure portal.

At around 0730 18/7/31 UTC we see some CPU spike ~50% on the VM running the windows node 21580k8s9000            
Again maybe unrelated

At around 0930 18/7/31 UTC is when the 
21580k8s9000  Pod sandbox changed, it will be killed and re-created.
Begins.

This morning we see that
 Normal  SandboxChanged  35s (x3601 over 14h)  kubelet, 21580k8s9000  Pod sandbox changed, it will be killed and re-created.
It has 2 restarts.

**What you expected to happen**:
pod to recover from sandbox changed.
Problem is I am not sure what is causing the sandbox changed, it could be our container. Container logs show the web server hosted therein was running OK up to the point of sandbox changed.

**How to reproduce it** (as minimally and precisely as possible):
Steps above and attached acs-engine config file and deployment yml.
Not sure if our image is required, maybe just a microsoft/dotnet-framework:4.7.2-runtime-20180615-windowsservercore-1709 based image will suffice?

**Anything else we need to know**:

We have seen this problem before with 1803 based windows images also, though it seemed to occur more readily.

Kubernetes.json (for acs-engine)
{
    "apiVersion": "vlabs",
    "properties": {
      "orchestratorProfile": {
        "orchestratorType": "Kubernetes",
        "orchestratorRelease": "1.11"
      },
      "masterProfile": {
        "count": 1,
        "dnsPrefix": "maclustermls2",
        "vmSize": "Standard_D2_v2"
      },
      "agentPoolProfiles": [        
        {
          "name": "winpool",
          "count": 1,
          "vmSize": "Standard_A2m_v2",
          "availabilityProfile": "AvailabilitySet",
          "osType": "Windows"
        }
      ],
      "windowsProfile": {
        "adminUsername": "--",
        "adminPassword": "--",
		"windowsPublisher": "MicrosoftWindowsServer",
		"windowsOffer": "WindowsServerSemiAnnual",
		"windowsSku": "Datacenter-Core-1709-with-Containers-smalldisk"
      },
      "linuxProfile": {
        "adminUsername": "--",
        "ssh": {
          "publicKeys": [
            {
              "keyData":  "--"
            }
          ]
        }
      },
      "servicePrincipalProfile": {
        "clientId": "--",
        "secret": "--"
      }
    }
}

Deployment.yml

apiVersion: apps/v1
kind: Deployment
metadata:
  name: symbolruntime
  labels:
    app: symbolruntime
spec:
  replicas: 1
  selector:
    matchLabels:
      app: symbolruntime
  template:
    metadata:
      labels:
        app: symbolruntime
    spec:
      containers:
      - name: symbolruntime
        image: cloudhmi.azurecr.io/pv:master-0.1.18199.6
        ports:
        - containerPort: 5000
      imagePullSecrets:
      - name: newsecret
      nodeSelector:
        beta.kubernetes.io/os: windows

apiVersion: v1
kind: Service
metadata:
  name: symbolruntime
  labels:
    app: symbolruntime
spec:
  type: LoadBalancer
  ports:
  - port: 80
    targetPort: 5000    
  selector:
    app: symbolruntime
  sessionAffinity: ClientIP 
[docker.log](https://github.com/Azure/acs-engine/files/2265364/docker.log)

[kubelet.zip](https://github.com/Azure/acs-engine/files/2265366/kubelet.zip)

**Comment by poirot007** [comment]
The pod went into an error state soon after it was created.
Events that happened:
1. Cluster was created
2. Single pod with an image based on windows 1709 was created.
3. The container was created successfully and it was running
4. Connected a browser to the external ip
5. Pod restarted. 
6. Reason was sandbox changed and it was killed and restarted.
7. Pod restarted twice before going into an error state.
8. Left the cluster to run over 2 days to check whether the pod would recover. But the pod had not recovered and when I did a kubectl pod described <pod name>, I got the following message : 
 _Normal  SandboxChanged  5s (x14657 over 3d)  kubelet, 42289k8s9000  Pod sandbox changed, it will be killed and re-created._

**Comment by poirot007** [comment]
The logs I got from the pod are : 
[docker.log](https://github.com/Azure/acs-engine/files/2264484/docker.log)
[kubelet.log](https://github.com/Azure/acs-engine/files/2264485/kubelet.log)

**Comment by poirot007** [comment]
I have seen the same problem mentioned in my previous comment. i.e the pod went into an error state soon after it was created.  The logs from the pods are : 
[docker.log](https://github.com/Azure/acs-engine/files/2265396/docker.log)
[kubelet.log](https://github.com/Azure/acs-engine/files/2265397/kubelet.log)

**Comment by PatrickLang** [comment]
Can you log into the Windows nodes directly and look for errors in `c:\k\kubelet.err.log` ? Those logs are attached from the Linux node, not the Windows node running the container.

Also - after the failure, what's the free disk space on c:? I'm wondering if the asp.net container logs are filling up the disk and causing failures.

**Comment by poirot007** [comment]
Hi Patrick,

When you say log into the windows nodes directly, do you mean remote desktop to the master VM and then get the err.log?  Do you have a set of steps that I can go through to get this log? Thanks for your help.

**Comment by poirot007** [comment]
Hi Patrick,
Have managed to get the kubelet.err.log : 
[kubelet.err-20180801T001658.151.log](https://github.com/Azure/acs-engine/files/2268940/kubelet.err-20180801T001658.151.log)

The free disk space on c:  12,117,385,216 bytes

**Comment by poirot007** [comment]
I have tried reproducing the scenario and again the pod went into an error state. The logs are : 
[logsFromPod_9-8-18.zip](https://github.com/Azure/acs-engine/files/2273132/logsFromPod_9-8-18.zip)

**Comment by poirot007** [comment]
I tried reproducing the scenario on a VM with better spec (Standard_DS3_V2) and again ran into the same problem.

**Comment by poirot007** [comment]
Have tried the same scenario with the following configurations : 
ACS Engine version : Acs-Engine 0.20.7
VM : DS3_V2
OSDisk Size : 2TB

Status : When I executed the same scenario, the pod restarted and the client wasn't able to display the graphics. The pod went into an error state after 6 restarts.

![image](https://user-images.githubusercontent.com/42057935/43947111-8dd10c34-9cca-11e8-9b98-c79d74287373.png)

**Kubernetes Version** : Client Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.2", GitCommit:"5fa2db2bd46ac79e5e00a4e6ed24191080aa463b", GitTreeState:"clean", BuildDate:"2018-01-18T10:09:24Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"windows/amd64"}
Server Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.2", GitCommit:"bb9ffb1654d4a729bb4cec18ff088eacc153c239", GitTreeState:"clean", BuildDate:"2018-08-07T23:08:19Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"linux/amd64"}

**The logs :** 
[DS3_V2_2TB_logs.zip](https://github.com/Azure/acs-engine/files/2277198/DS3_V2_2TB_logs.zip)

**Azure CNI plugin version and configuration :**
mode : "bridge"
dns.Nameservers : "10.0.0.10",
"168.63.129.16"
dns.Search : "svc.cluster.local"

**Azure CNI build** : v1.0.7

**Comment by jasonshortphd** [comment]
@PatrickLang any theory on things they can keep trying?  
@poirot007 any way to build a smaller in scope sample to repro the issue?  Trying different containers, or different apps on the container?

**Comment by PatrickLang** [comment]
I do see some network failures that weren't handled correctly. ACS-Engine 0.20.9 fixes some known errors and also fixes error handling https://github.com/Azure/acs-engine/pull/3667. It should be out later today. It would be worth redeploying with that release as the next step in the investigations.

The other thing I'm doing is adding some more logs to my [log collection tool](https://github.com/patricklang/logslurp) so we can see if there are other sources of errors buried in the Windows event logs such as `Microsoft-Windows-Hyper-V-Compute/Admin`

**Comment by poirot007** [comment]
@jasonshortphd  
As suggested by @paulbouwer and @PatrickLang  , I will test the following 2 sets of configurations and then update the status of this issue : 
**Configuration 1 :** 
ACS Engine (0.20.9)
VM : DS3V2
OSDiskSize : 1023GB
Windows 1709 Image

**Configuration 2** 
ACS Engine (0.20.9)
VM : DS3V2
OSDiskSize : 1023GB
Windows 1803 Image

**Comment by poirot007** [comment]
Tested with the following configuration : 
Configuration 1 :
ACS Engine (0.20.9)
VM : DS3V2
OSDiskSize : 1023GB
Windows 1709 Image

When I connected 1 client, the pod restarted and went into an error state and then immediately recovered. I then connected a few more clients and the pod again went into an error state and then recovered.
Screenshot : 
![image](https://user-images.githubusercontent.com/42057935/44069835-1b960fb4-9fc4-11e8-862e-5c39aff7bf2a.png)

Everytime a new client is added, the pod goes into an error state and then recovers.

**Logs:** 
[logs.zip](https://github.com/Azure/acs-engine/files/2285446/logs.zip)

**Azure CNI plugin version and configuration :**
mode : "bridge"
dns.Nameservers : "10.0.0.10",
"168.63.129.16"
dns.Search : "svc.cluster.local"

**Azure CNI build :** v1.0.11

Kubernetes version : 
Client Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.2", GitCommit:"5fa2db2bd46ac79e5e00a4e6ed24191080aa463b", GitTreeState:"clean", BuildDate:"2018-01-18T10:09:24Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"windows/amd64"}
Server Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.2", GitCommit:"bb9ffb1654d4a729bb4cec18ff088eacc153c239", GitTreeState:"clean", BuildDate:"2018-08-07T23:08:19Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"linux/amd64"}
